src/ray/core_worker/core_worker.cc

```diff
@@ -1444,69 +1444,102 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
-  if (!borrowed_ids.empty()) {
-    reference_counter_->GetAndClearLocalBorrowers(borrowed_ids, borrowed_refs);
-  }
-  // Unpin the borrowed IDs.
-  std::vector<ObjectID> deleted;
-  for (const auto &borrowed_id : borrowed_ids) {
-    RAY_LOG(DEBUG) << "Decrementing ref for borrowed ID " << borrowed_id;
-    reference_counter_->RemoveLocalReference(borrowed_id, &deleted);
-  }
-  if (options_.ref_counting_enabled) {
-    memory_store_->Delete(deleted);
-  }
+    if (task_spec.IsNormalTask() && reference_counter_->NumObjectIDsInScope() != 0) {
+      RAY_LOG(DEBUG)
+          << "There were " << reference_counter_->NumObjectIDsInScope()
+          << " ObjectIDs left in scope after executing task " << task_spec.TaskId()
+          << ". This is either caused by keeping references to ObjectIDs in Python "
+             "between "
+             "tasks (e.g., in global variables) or indicates a problem with Ray's "
+             "reference counting, and may cause problems in the object store.";
+    }
 
-  if (task_spec.IsNormalTask() && reference_counter_->NumObjectIDsInScope() != 0) {
-    RAY_LOG(DEBUG)
-        << "There were " << reference_counter_->NumObjectIDsInScope()
-        << " ObjectIDs left in scope after executing task " << task_spec.TaskId()
-        << ". This is either caused by keeping references to ObjectIDs in Python "
-           "between "
-           "tasks (e.g., in global variables) or indicates a problem with Ray's "
-           "reference counting, and may cause problems in the object store.";
-  }
+    if (!options_.is_local_mode) {
+      SetCurrentTaskId(TaskID::Nil());
+      worker_context_.ResetCurrentTask(task_spec);
+    }
+    {
+      absl::MutexLock lock(&mutex_);
+      current_task_ = TaskSpecification();
+    }
+    RAY_LOG(DEBUG) << "Finished executing task " << task_spec.TaskId();
+
+    if (status.IsSystemExit()) {
+      Exit(status.IsIntentionalSystemExit());
+    }
+
+    // SANG-TODO: Call receiver->run_accept_callback_on_completion(taskID)
+    if (!options_.is_local_mode) {
+      // Local Mode doesn't require accept callback.
+      direct_task_receiver_->CompleteAcceptRequest(task_spec.TaskId(), *return_objects,
+                                                   status);
+    }
+  };
 
-  if (!options_.is_local_mode) {
-    SetCurrentTaskId(TaskID::Nil());
-    worker_context_.ResetCurrentTask(task_spec);
-  }
   {
     absl::MutexLock lock(&mutex_);
-    current_task_ = TaskSpecification();
+    // SAGN-TODO Add completion callback so that it can be used to finish up the task.
+    on_excute_task_completion_map.emplace(task_spec.TaskId(), on_execute_task_completion);
   }
-  RAY_LOG(DEBUG) << "Finished executing task " << task_spec.TaskId();
+  status = options_.task_execution_callback(
+      task_type, func, task_spec.GetRequiredResources().GetResourceMap(), args,
+      arg_reference_ids, return_ids, return_objects);
+  OnExecuteTaskCompletion(task_spec.TaskId(), status);
```

Comment:
This should only be called if sync, right?

Reply:
yes. Code here only contains refactoring.